File: web_version/pmugg_files/shape/vec3.js
// 3D Vectors
ms.vec3 = function(x, y, z) {
	this.x = x;
	this.y = y;
	this.z = z;
};

ms.vec3.ORIGIN = new ms.vec3(0, 0, 0);
ms.vec3.X_HAT = new ms.vec3(1, 0, 0);
ms.vec3.Y_HAT = new ms.vec3(0, 1, 0);
ms.vec3.Z_HAT = new ms.vec3(0, 0, 1);

ms.vec3.create = function(v) {
	return new ms.vec3(v.x, v.y, v.z);
};

ms.vec3.prototype.copy = function() {
	return new ms.vec3(this.x, this.y, this.z);
};

ms.vec3.prototype.toArray = function() {
	return [this.x, this.y, this.z];
};

// Copy from a vector.
ms.vec3.prototype.set = function(a) {
  this.x = a.x;
  this.y = a.y;
  this.z = a.z;
};

ms.vec3.prototype.set3 = function(x, y, z) {
  this.x = x;
  this.y = y;
  this.z = z;
};

ms.vec3.prototype.distance2 = function(position) {
	var dx = this.x - position.x;
	var dy = this.y - position.y;
	var dz = this.z - position.z;
	return dx * dx + dy * dy + dz * dz;
};

ms.vec3.prototype.distance = function(position) {
	return Math.sqrt(this.distance2(position));
};

ms.vec3.prototype.length = function() {
	return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
};

ms.vec3.prototype.length2 = function(position) {
	return this.x * this.x + this.y * this.y  + this.z * this.z;
};

ms.vec3.prototype.add = function(position) {
	this.x += position.x;
	this.y += position.y;
	this.z += position.z;
	return this;
};

ms.vec3.prototype.minus = function(position) {
	this.x -= position.x;
	this.y -= position.y;
	this.z -= position.z;
	return this;
};

ms.vec3.prototype.normalize = function() {
	var r2 = this.x * this.x + this.y * this.y + this.z * this.z;
	if (r2 != 0) {
		this.scale(1 / Math.sqrt(r2));
	}
	return this;
};

ms.vec3.prototype.scale = function(s) {
	this.x *= s;
	this.y *= s;
	this.z *= s;
	return this;
};

ms.vec3.lerp = function(start, end, s) {
	var x = (1 - s) * start.x + s * end.x;
	var y = (1 - s) * start.y + s * end.y;
	var z = (1 - s) * start.z + s * end.z;
	return new ms.vec3(x, y, z);
};

// Compute the dot product.
ms.vec3.prototype.dot = function(v) {
	return this.x * v.x + this.y * v.y + this.z * v.z;
};

ms.vec3.prototype.print = function(v) {
	return '(' + parseFloat(this.x).toFixed(3) + ', ' + parseFloat(this.y).toFixed(3) + ', ' + parseFloat(this.z).toFixed(3) + ')';
};

ms.vec3.prototype.cross = function(b) {
	var a = this;
	return new ms.vec3(
		a.y * b.z - a.z * b.y,
		a.z * b.x - a.x * b.z,
		a.x * b.y - a.y * b.x);
};

// Magnitude of the cross product in the z-direction.
ms.vec3.prototype.crossZ = function(v) {
  return this.x * v.y - this.y * v.x;
};

// Get the value along a particular dimension.
ms.vec3.prototype.getValue = function(dim) {
  if (dim == 0) {
    return this.x;
  } else if (dim == 1) {
    return this.y; 
  } else if (dim == 2) {
    return this.z; 
  } else {
    alert('Get value undefined for dimension' + dim);
  }
};

// Get the value along a particular dimension.
ms.vec3.prototype.setValue = function(value, dim) {
  if (dim == 0) {
    this.x = value;
  } else if (dim == 1) {
    this.y = value;
  } else {
    this.z = value;	  
  }
};

// Drop one of the dimensions. Returns a vec2 with it missing.
ms.vec3.prototype.dropDim = function(dim) {
	if (dim == 0) {
		return new ms.vec2(this.y, this.z);
	} else if (dim == 1) {
		return new ms.vec2(this.x, this.z);
	} else {
		return new ms.vec2(this.x, this.y);
	}
};


// Determines the xy lexicographical order of two points for sweepline.
/* ms.vec3.prototype.compare = function(other_point){
  // x-coord first
  if (this.x > other_point.x) return  1; 
  if (this.x < other_point.x) return -1;

  // y-coord second
  if (this.y > other_point.y) return  1; 
  if (this.y < other_point.y) return -1;

  // they are the same point
  return 0;  
}; */

ms.vec3.rotationMatrix = function(theta) {
	var cosTheta = Math.cos(theta);
	var sinTheta = Math.sin(theta);
	return [[cosTheta, -sinTheta, 0], [sinTheta, cosTheta, 0], [0, 0, 1]];
};

ms.vec3.prototype.matrixMultiply = function(A) {
	var x = A[0][0] * this.x + A[0][1] * this.y + A[0][2] * this.z;
	var y = A[1][0] * this.x + A[1][1] * this.y + A[1][2] * this.z;
	var z = A[2][0] * this.x + A[2][1] * this.y + A[2][2] * this.z;
	this.x = x;
	this.y = y;
	this.z = y;
};

ms.vec3.prototype.rotate = function(theta) {
	var R = ms.vec3.rotationMatrix(theta);
	this.matrixMultiply(R);
	return this;
};

/*
// tests if point is Left|On|Right of the line P0 to P1.
//
// returns: 
//  >0 for left of the line 
//  0 for on the line
//  <0 for right of the line
ms.vec3.prototype.is_left = function(p0, p1){
 return (p1.x - p0.x) * (this.y - p0.y) - (this.x - p0.x) * (p1.y - p0.y);  
} */

ms.vec3.coordinatesClose = function(a, b, tolerance) {
	return (Math.abs(a.x - b.x) <= tolerance &
		Math.abs(a.y - b.y) <= tolerance & 
		Math.abs(a.z - b.z) <= tolerance);
};

ms.vec3.prototype.coordinatesClose = function(b, tolerance) {
	return ms.vec3.coordinatesClose(this, b, tolerance);
};

ms.vec3.unitVec = function(angle) {
	return new ms.vec3(Math.cos(angle), Math.sin(angle), 0);
};

/*
ms.vec3.angle = function(startPos, endPos) {
	var dx = endPos.x - startPos.x;
	var dy = endPos.y - startPos.y;
	return Math.atan2(dy, dx);
}; */

ms.vec3.prototype.move = function(dx, dy, dz) {
	this.x += dx;
	this.y += dy;
	this.z += dz;
};

// Convert from a vector to a matrix.
ms.vec3.prototype.toMatrix = function() {
	return mathG.matrix([[this.x], [this.y], [this.z]]);
};

// Return two vectors orthonormal to this one.
ms.vec3.prototype.orthonormal = function() {
	var q = new ms.vec3(1, 0, 0);
	if (Math.abs(this.dot(q)) > 0.9) {
		q = new ms.vec3(0, 1, 0);
	}
	var v = this.cross(q);
	var u = v.cross(this);
	u.normalize();
	v.normalize();
	return {u, v};
};

// The angle axis required to rotate one vector to parallel to the other one.
ms.vec3.angleAxis = function(p, q) {
	var axis = p.cross(q);
	if (Math.abs(axis.length2) < 1e-6) {
		return {axis: new ms.vec3(0, 0, 1), angle: 0};
	}
	axis.normalize();
	// The p, ay, and axis would be x, y, and z if p and q were on the z = 0 plane.
	ay = axis.cross(p);
	ay.normalize();

	var angle = Math.atan2(ay.dot(q), p.dot(q));
	return {axis, angle: 180 / Math.PI * angle};
};

// Swap axes going from Y is up to Z is up.
ms.vec3.prototype.swapAxes = function() {
	return new ms.vec3(this.z, this.x, this.y);
};

ms.vec3.prototype.export = function() {
	return { x: this.x, y: this.y, z: this.z };
};

ms.vec3.EPS = 1e-5;

ms.vec3.import = function(json) {
	var result = new ms.vec3(json.x, json.y, json.z);
	// Round very small numbers to 0. This fixes an issue with computing the turns.
	// TODO: Make computing turns more robust.
	for (var i = 0; i < 3; i++) {
		if (Math.abs(result.getValue(i)) < ms.vec3.EPS) {
			result.setValue(0, i);
		}
	}
	result.normalize();
	return result;
};

File: cpp_version/pmugg/shape/vec3.cpp
#include "pch.h"
#include "vec3.h"
#include "vec2.h"
#include <cmath>
#include <sstream>
#include <iomanip>

namespace ms {

const Vec3 Vec3::ORIGIN(0, 0, 0);
const Vec3 Vec3::X_AXIS(1, 0, 0);
const Vec3 Vec3::Y_AXIS(0, 1, 0);
const Vec3 Vec3::Z_AXIS(0, 0, 1);

Vec3::Vec3(float x_, float y_, float z_)
    : x(x_)
    , y(y_)
    , z(z_) {}

float Vec3::getValue(int dim) const {
    switch (dim) {
        case 0: return x;
        case 1: return y;
        case 2: return z;
        default: throw std::runtime_error("Invalid dimension for Vec3");
    }
}

void Vec3::setValue(float value, int dim) {
    switch (dim) {
        case 0: x = value; break;
        case 1: y = value; break;
        case 2: z = value; break;
        default: throw std::runtime_error("Invalid dimension for Vec3");
    }
}

Vec3& Vec3::add(const Vec3& v) {
    x += v.x;
    y += v.y;
    z += v.z;
    return *this;
}

Vec3& Vec3::minus(const Vec3& v) {
    x -= v.x;
    y -= v.y;
    z -= v.z;
    return *this;
}

Vec3& Vec3::scale(float s) {
    x *= s;
    y *= s;
    z *= s;
    return *this;
}

Vec3& Vec3::normalize() {
    float len = length();
    if (len != 0) {
        scale(1.0f / len);
    }
    return *this;
}

float Vec3::dot(const Vec3& v) const {
    return x * v.x + y * v.y + z * v.z;
}

Vec3 Vec3::cross(const Vec3& v) const {
    return Vec3(
        y * v.z - z * v.y,
        z * v.x - x * v.z,
        x * v.y - y * v.x
    );
}

float Vec3::length() const {
    return std::sqrt(length2());
}

float Vec3::length2() const {
    return x * x + y * y + z * z;
}

float Vec3::distance(const Vec3& v) const {
    return std::sqrt(distance2(v));
}

float Vec3::distance2(const Vec3& v) const {
    float dx = x - v.x;
    float dy = y - v.y;
    float dz = z - v.z;
    return dx * dx + dy * dy + dz * dz;
}

Vec3& Vec3::rotate(const Vec3& axis, float theta) {
    float cosTheta = std::cos(theta);
    float sinTheta = std::sin(theta);
    float oneMinusCosTheta = 1.0f - cosTheta;

    float ux = axis.x;
    float uy = axis.y;
    float uz = axis.z;

    float newX = (cosTheta + ux * ux * oneMinusCosTheta) * x +
                 (ux * uy * oneMinusCosTheta - uz * sinTheta) * y +
                 (ux * uz * oneMinusCosTheta + uy * sinTheta) * z;

    float newY = (uy * ux * oneMinusCosTheta + uz * sinTheta) * x +
                 (cosTheta + uy * uy * oneMinusCosTheta) * y +
                 (uy * uz * oneMinusCosTheta - ux * sinTheta) * z;

    float newZ = (uz * ux * oneMinusCosTheta - uy * sinTheta) * x +
                 (uz * uy * oneMinusCosTheta + ux * sinTheta) * y +
                 (cosTheta + uz * uz * oneMinusCosTheta) * z;

    x = newX;
    y = newY;
    z = newZ;
    return *this;
}

Vec3& Vec3::rotateX(float theta) {
    float cosTheta = std::cos(theta);
    float sinTheta = std::sin(theta);
    float newY = y * cosTheta - z * sinTheta;
    float newZ = y * sinTheta + z * cosTheta;
    y = newY;
    z = newZ;
    return *this;
}

Vec3& Vec3::rotateY(float theta) {
    float cosTheta = std::cos(theta);
    float sinTheta = std::sin(theta);
    float newX = x * cosTheta + z * sinTheta;
    float newZ = -x * sinTheta + z * cosTheta;
    x = newX;
    z = newZ;
    return *this;
}

Vec3& Vec3::rotateZ(float theta) {
    float cosTheta = std::cos(theta);
    float sinTheta = std::sin(theta);
    float newX = x * cosTheta - y * sinTheta;
    float newY = x * sinTheta + y * cosTheta;
    x = newX;
    y = newY;
    return *this;
}

Vec3 Vec3::copy() const {
    return *this;
}

Vec2 Vec3::dropDim(int dim) const {
    if (dim == 0) {
        return Vec2(y, z);
    } else if (dim == 1) {
        return Vec2(x, z);
    } else {
        return Vec2(x, y);
    }
}

std::string Vec3::toString() const {
    std::stringstream ss;
    ss << std::fixed << std::setprecision(3);
    ss << "(" << x << ", " << y << ", " << z << ")";
    return ss.str();
}

Vec3 Vec3::lerp(const Vec3& start, const Vec3& end, float s) {
    return Vec3(
        (1 - s) * start.x + s * end.x,
        (1 - s) * start.y + s * end.y,
        (1 - s) * start.z + s * end.z
    );
}

Vec3 Vec3::unitVec(float theta, float phi) {
    float sinPhi = std::sin(phi);
    return Vec3(
        std::cos(theta) * sinPhi,
        std::sin(theta) * sinPhi,
        std::cos(phi)
    );
}

bool Vec3::coordinatesClose(const Vec3& a, const Vec3& b, float tolerance) {
    return (std::abs(a.x - b.x) <= tolerance && 
            std::abs(a.y - b.y) <= tolerance &&
            std::abs(a.z - b.z) <= tolerance);
}

Vec3 Vec3::swapAxes() const {
    return Vec3(z, x, y);
}

Vec3 Vec3::import(const Json& json) {
    if (json.is_array()) {
        return Vec3(json[0], json[1], json[2]);
    } else {
        return Vec3(json["x"], json["y"], json["z"]);
    }
}

} // namespace ms 